src/meta/GflagsManager.h

```diff
@@ -0,0 +1,127 @@
+
+public:
+    // methods for consoles, reg/set/get/list configs from meta server
+    virtual folly::Future<StatusOr<bool>> setConfig(const cpp2::ConfigModule& module,
+                                                    const std::string& name,
+                                                    const cpp2::ConfigType& type,
+                                                    const VariantType& value) = 0;
+
+    virtual folly::Future<StatusOr<std::vector<cpp2::ConfigItem>>>
+    getConfig(const cpp2::ConfigModule& module, const std::string& name) = 0;
+
+    virtual folly::Future<StatusOr<std::vector<cpp2::ConfigItem>>>
+    listConfigs(const cpp2::ConfigModule& module) = 0;
+
+    virtual folly::Future<StatusOr<bool>> registerConfig(const cpp2::ConfigModule& module,
+                                                         const std::string& name,
+                                                         const cpp2::ConfigType& type,
+                                                         const cpp2::ConfigMode& mode,
+                                                         const std::string& defaultValue) = 0;
+
+    // check the config is registered in cache or not
+    Status isCfgRegistered(const cpp2::ConfigModule& module,
+                           const std::string& name, const cpp2::ConfigType& type);
+
+    // methods for code to get config from cache
+    StatusOr<int64_t> getConfigAsInt64(const std::string& name);
+
+    StatusOr<double> getConfigAsDouble(const std::string& name);
+
+    StatusOr<bool> getConfigAsBool(const std::string& name);
+
+    StatusOr<std::string> getConfigAsString(const std::string& name);
+
+protected:
+    virtual ~GflagsManager();
+    virtual void loadCfgThreadFunc() = 0;
+    virtual void regCfgThreadFunc() = 0;
+
+    void init();
+
+    void declareGflags();
+
+    StatusOr<ConfigItem> getConfigFromCache(const cpp2::ConfigModule& module,
+                                            const std::string& name,
+                                            const cpp2::ConfigType& type);
+
+    // declare is for using GflagsManager in code, We don't check whether config cache ready or not.
+    // We just put default value in cache, and when we load configs stored in meta successfully, it
+    // would replaced by value from meta.
+    Status declareConfig(const cpp2::ConfigModule& module, const std::string& name,
+                         const cpp2::ConfigType& type, const cpp2::ConfigMode& mode,
+                         const VariantType& defaultValue);
+
+    void updateConfigCache(const std::vector<cpp2::ConfigItem>& items);
+
+    void updateGflagsValue(const ConfigItem& item);
+
+    template<typename ValueType>
+    std::string gflagsValueToThriftValue(const gflags::CommandLineFlagInfo& flag);
+
+    cpp2::ConfigModule                  module_{cpp2::ConfigModule::UNKNOWN};
+    std::vector<cpp2::ConfigItem>       gflagsDeclared_;
+    thread::GenericWorker               bgThread_;
```

Comment:
Do we need the bgThread in base class?

Reply:
Yeah, I will move it to ClientBasedGflagsManager